crates/messages/src/assignments.rs

```diff
@@ -0,0 +1,346 @@
+use core::str;
+use std::collections::HashMap;
+#[cfg(feature = "assignment_reader")]
+use std::collections::{BTreeMap, VecDeque};
+
+use anyhow::anyhow;
+#[cfg(feature = "assignment_writer")]
+use crypto_box::aead::{AeadCore, OsRng};
+use crypto_box::{aead::Aead, PublicKey, SalsaBox, SecretKey};
+#[cfg(feature = "assignment_writer")]
+use curve25519_dalek::edwards::CompressedEdwardsY;
+#[cfg(feature = "assignment_reader")]
+use flate2::read::GzDecoder;
+#[cfg(feature = "assignment_writer")]
+use log::error;
+use serde::{Deserialize, Serialize};
+#[cfg(feature = "assignment_reader")]
+use serde_json::Value;
+use serde_with::base64::Base64;
+use serde_with::serde_as;
+#[cfg(feature = "assignment_reader")]
+use sha2::Digest;
+#[cfg(feature = "assignment_reader")]
+use sha2::Sha512;
+#[cfg(feature = "assignment_reader")]
+use sha3::digest::generic_array::GenericArray;
+
+#[cfg(feature = "assignment_writer")]
+use base64::{engine::general_purpose::STANDARD as base64, Engine};
+#[cfg(feature = "assignment_writer")]
+use hmac::{Hmac, Mac};
+#[cfg(feature = "assignment_writer")]
+use sha2::Sha256;
+#[cfg(feature = "assignment_writer")]
+use url::form_urlencoded;
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct Chunk {
+    pub id: String,
+    pub base_url: String,
+    pub files: HashMap<String, String>,
+    pub size_bytes: u64,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Dataset {
+    pub id: String,
+    pub base_url: String,
+    pub chunks: Vec<Chunk>,
+}
+
+#[derive(Debug, Default, Serialize, Deserialize)]
+#[serde(rename_all = "kebab-case")]
+struct Headers {
+    worker_id: String,
+    worker_signature: String,
+}
+
+#[serde_as]
+#[derive(Debug, Default, Serialize, Deserialize, Clone)]
+#[serde(rename_all = "camelCase")]
+struct EncryptedHeaders {
+    #[serde_as(as = "Base64")]
+    identity: Vec<u8>,
+    #[serde_as(as = "Base64")]
+    nonce: Vec<u8>,
+    #[serde_as(as = "Base64")]
+    ciphertext: Vec<u8>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct WorkerAssignment {
+    status: String,
+    chunks_deltas: Vec<u64>,
+    encrypted_headers: EncryptedHeaders,
+}
+
+#[derive(Debug, Default, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Assignment {
+    datasets: Vec<Dataset>,
+    worker_assignments: HashMap<String, WorkerAssignment>,
+    #[serde(skip)]
```

Comment:
For completeness, we can avoid warning for `assignment_reader` builds like this:
```suggestion
    #[cfg(feature = "assignment_writer")]
    #[serde(skip)]
```